core/src/main/java/com/linecorp/armeria/common/util/DefaultAsyncLoader.java

```diff
@@ -0,0 +1,268 @@
+                handleException.handle((val, cause) -> {
+                    if (cause != null) {
+                        logger.warn("Failed to load a new value from exceptionHandler: {}." +
+                                    "the previous value: {}", exceptionHandler, cache, cause);
+                        future.completeExceptionally(cause);
+                    } else {
+                        future.complete(new CacheEntry<>(val));
+                    }
+                    return null;
+                });
+            } else {
+                future.completeExceptionally(originCause);
+            }
+        } catch (Exception e) {
+            logger.warn("Unexpected exception from exceptionHandler.apply()", e);
+            future.completeExceptionally(originCause);
+        }
+    }
+
+    private boolean isValid(@Nullable CacheEntry<T> cacheEntry) {
+        if (cacheEntry == null) {
+            return false;
+        }
+
+        if (expireAfterLoad != null) {
+            final Instant expiration = cacheEntry.loadWhen.plusMillis(expireAfterLoad.toMillis());
+            if (Instant.now().isAfter(expiration)) {
+                return false;
+            }
+        }
+
+        if (expireIf != null && expireIf.test(cacheEntry.loadVal)) {
+            return false;
+        }
+
+        return true;
+    }
+
+    private static class CacheEntry<T> {
+
+        private final T loadVal;
+        private final Instant loadWhen = Instant.now();
+
+        CacheEntry(T loadVal) {
+            requireNonNull(loadVal, "loadVal");
+            this.loadVal = loadVal;
+        }
+    }
+
+    private static class RefreshingFuture<U> extends CompletableFuture<CacheEntry<U>> {
+
+        public static <T> RefreshingFuture<T> completedFuture() {
+            final RefreshingFuture<T> future = new RefreshingFuture<>(null);
+            future.complete(null);
+            return future;
+        }
+
+        @Nullable
+        private volatile CacheEntry<U> cacheEntry;
+        // True when refreshing is in progress, means not yet completed by new value or exception.
+        private volatile boolean refreshing;
+
+        RefreshingFuture(@Nullable CacheEntry<U> cacheEntry) {
+            // cacheEntry can be null if it expired or has not loaded yet.
+            this.cacheEntry = cacheEntry;
+            if (cacheEntry != null) {
+                complete(cacheEntry);
+            }
+            refreshing = true;
+        }
+
+        @Override
+        public boolean complete(CacheEntry<U> newVal) {
+            refreshing = false;
+            cacheEntry = newVal;
+            obtrudeValue(newVal);
+            return true;
+        }
+
+        @Override
+        public boolean completeExceptionally(Throwable ex) {
+            refreshing = false;
+            if (cacheEntry != null) {
+                logger.warn("Failed to refresh a new value. the previous value: {}", cacheEntry.loadVal, ex);
```

Comment:
```
1. When cached: 1, expireAfterLoad: 10sec, refresh every 1 sec
2. After 1 sec, try to refresh it
3. If refresh failed, we expect to get `cached: 1`, not exception! // 👈 👈
```
If `cacheEntry != null`, it means something is cached at first and we're trying to `refresh` it.

So if refreshing failed here, I want to return cached value instead of throwing exception!

If we `obtrudeException(ex)` on L252 too, `get()` return exception after refreshing failure and on next `get()`, we load new value although cached value is not expired yet.

Reply:
Make sense.